### ISSUE TITLE
Node searcher fails in `RuboCop::AST::Node`

### ISSUE THREAD

**Issue by thomthom** [bbatsov/rubocop#5163]
I'm creating a gem with custom RuboCop rules: https://github.com/SketchUp/rubocop-sketchup

One of the cops is running into an error while running the following AST searcher:

```ruby
def_node_search :force_encoding, <<-PATTERN
  (send
    (_ $_)
    :force_encoding ...)
PATTERN
```

I've condensed the triggering condition to the following test:

```ruby
  it 'does not raise an error with module attr_accessor' do
    inspect_source(['module Example',
                    '  class << self; attr_accessor :foo; end',
                    '  file = __FILE__',
                    'end'])
    expect(cop.offenses.size).to eq(1)
  end
```

When I run the cop in a production environment with `--debug` flag I get the following backtrace:

```bash
An error occurred while SketchupSuggestions/FileEncoding cop was inspecting C:/Users/tthomas2/SourceTree/ExtensionReviewTools/static-analysis/temp/curic_detail_view/loader.rb:3:1.
undefined method `children' for nil:NilClass
C:3:in `block in force_encoding'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:523:in `block in visit_descendants'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:173:in `block in each_child_node'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:171:in `each'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:171:in `each_child_node'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:522:in `visit_descendants'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:524:in `block in visit_descendants'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:173:in `block in each_child_node'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:171:in `each'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:171:in `each_child_node'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:522:in `visit_descendants'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/node.rb:248:in `each_node'
C:2:in `force_encoding'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-sketchup-0.1.1/lib/rubocop/sketchup/suggestions/file_encoding.rb:67:in `each'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-sketchup-0.1.1/lib/rubocop/sketchup/suggestions/file_encoding.rb:67:in `to_a'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-sketchup-0.1.1/lib/rubocop/sketchup/suggestions/file_encoding.rb:67:in `on_assign'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/commissioner.rb:44:in `block (2 levels) in on_lvasgn'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/commissioner.rb:109:in `with_cop_error_handling'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/commissioner.rb:43:in `block in on_lvasgn'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/commissioner.rb:42:in `each'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/commissioner.rb:42:in `on_lvasgn'
(eval):2:in `block in on_begin'
(eval):2:in `each'
(eval):2:in `on_begin'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/commissioner.rb:48:in `on_begin'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/traversal.rb:141:in `on_while'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/commissioner.rb:48:in `on_module'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/ast/traversal.rb:12:in `walk'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/commissioner.rb:60:in `investigate'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/team.rb:114:in `investigate'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/team.rb:102:in `offenses'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cop/team.rb:44:in `inspect_file'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:258:in `inspect_file'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:205:in `block in do_inspection_loop'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:237:in `block in iterate_until_no_changes'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:230:in `loop'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:230:in `iterate_until_no_changes'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:201:in `do_inspection_loop'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:111:in `block in file_offenses'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:121:in `file_offense_cache'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:109:in `file_offenses'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:100:in `process_file'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:78:in `block in each_inspected_file'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:75:in `each'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:75:in `reduce'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:75:in `each_inspected_file'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:67:in `inspect_files'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/runner.rb:39:in `run'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cli.rb:87:in `execute_runner'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/lib/rubocop/cli.rb:32:in `run'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/bin/rubocop:13:in `block in <top (required)>'
C:/Ruby23/lib/ruby/2.3.0/benchmark.rb:308:in `realtime'
C:/Ruby23/lib/ruby/gems/2.3.0/gems/rubocop-0.51.0/bin/rubocop:12:in `<top (required)>'
C:/Ruby23/bin/rubocop:22:in `load'
C:/Ruby23/bin/rubocop:22:in `<main>'
```

I've tried tracking it down, but I get lost at `rubocop/lib/ast/node.rb:171` (RuboCop 0.51):

```ruby
      def each_child_node(*types)
        return to_enum(__method__, *types) unless block_given?

        children.each do |child|
          next unless child.is_a?(Node)
          yield child if types.empty? || types.include?(child.type)
        end

        self
      end
```

The error indicate the call to `children` is the source - but I'm not able to find the source for that method/accessor. Nor did the debugger let me step into it.

Is `children` coming from `Parser::AST`? Any ideas on how to debug what's going on there?

--------

## Expected behavior

Expected that the node matcher does't raise any errors while inspecting the source.

## Actual behavior

RuboCop fails with the following error when running the spec:

```bash
Failures:

  1) RuboCop::Cop::SketchupSuggestions::FileEncoding does not raise an error with module attr_accessor
     Failure/Error: encoded = force_encoding(node.parent).to_a

     NoMethodError:
       undefined method `children' for nil:NilClass
     # C:3:in `block in force_encoding'
     # C:2:in `force_encoding'
     # ./lib/rubocop/sketchup/suggestions/file_encoding.rb:59:in `each'
     # ./lib/rubocop/sketchup/suggestions/file_encoding.rb:59:in `to_a'
     # ./lib/rubocop/sketchup/suggestions/file_encoding.rb:59:in `on_assign'
     # (eval):2:in `block in on_begin'
     # (eval):2:in `each'
     # (eval):2:in `on_begin'
     # ./spec/rubocop/sketchup/suggestions/file_encoding_spec.rb:63:in `block (2 levels) in <top (required)>'
```

## Steps to reproduce the problem

1. clone https://github.com/SketchUp/rubocop-sketchup.git
2. checkout `dev-file-encoding-bug`
3. `bundle install`
4. `bundle exec rake spec SPEC=spec/rubocop/sketchup/suggestions/file_encoding_spec.rb:62`

`file = __FILE__` will trigger `on_assign` in the cop on [line 53](https://github.com/SketchUp/rubocop-sketchup/blob/dev-file-encoding-bug/lib/rubocop/sketchup/suggestions/file_encoding.rb#L53)

It then takes the parent node for that assignment match and run the node searcher.

## RuboCop version

```
$ rubocop -V
0.51.0 (using Parser 2.4.0.2, running on ruby 2.3.3 i386-mingw32)
```

**Comment by Drenmi** [comment]
Without trying it out, the problem is most likely with how you attempt to match the receiver:

```
(send (_ $_) :force_encoding ...)
```

The `(_ $_)` looks very strange. It doesn't say what node type you're trying to match. 🙂 In `attr_accessor :foo`, the receiver is `nil`, but you try to enquire on the children which is probably why you get the `NoMethodError`.

I suspect from your test file that maybe you meant:

```
(send (send _ $_) :force_encoding ...)
```

**Comment by thomthom** [comment]
Yea, I am not surprised that my syntax might be unusual. Only recently starting working with these cops.

The `attr_accessor ` isn't actually what I'm looking for - it's simply code that triggers the error to be thrown.

The I'm looking for is stuff like this:

```ruby
file = __FILE__
path = File.dirname(file)
```

A little background here: I work for SketchUp, a 3d modelling application, that embeds a Ruby interpreter into it's desktop client. We expose a Ruby API to allow people to write extensions for out application. A common run-time error is using __FILE__ which under Windows will have the wrong encoding if the path contain Unicode characters.

So my strategy is to catch usage of __FILE__ and warn about it. If it's used directly in a method call we yield a warning. For this I use `on_send` and it's working fine. (https://github.com/SketchUp/rubocop-sketchup/blob/dev-file-encoding-bug/lib/rubocop/sketchup/suggestions/file_encoding.rb#L40)

But I'm also trying to do some basic checks on `__FILE__` being applied to a variable. A workaround for the encoding issue with `__FILE__` is to use `.force_encoding` to correct it. So my strategy is to detect when a variable/constant is being assigned the value of `__FILE__`. Upon such an assignment I then check the parent node to see if `.force_encoding` is called upon this variable.

But it could be assigned to a local variable, instance variable, class variable, global variable or constant. So I'm trying to catch all of that. Hence was being broad in the type I was trying to catch.

In my attempt to do that I copied a pattern from another cop, by aliasing the various assignment callback:

```ruby
        def on_assign(node)
          lhs, value = *node
          return unless magic_file_or_dir?(value)
          # After assigning __FILE__ or __dir_ to a variable, check the parent
          # scope to whether .force_encoding is called on the variable.
          return if node.parent.nil?
          encoded = force_encoding(node.parent).to_a
          return if encoded.include?(lhs)
          add_offense(node)
        end

        # TODO: Review this list of aliases.
        alias on_lvasgn   on_assign
        alias on_masgn    on_assign
        alias on_casgn    on_assign
        alias on_ivasgn   on_assign
        alias on_cvasgn   on_assign
        alias on_gvasgn   on_assign
        alias on_or_asgn  on_assign
        alias on_and_asgn on_assign
        alias on_op_asgn  on_assign
```

I don't actually know if I really need all of these aliases. But, I can reproduce this error with only `on_lvasgn`.

In summary, for stuff like the following my matcher works fine:

```ruby
file = __FILE__
path = File.dirname(file)
```

```ruby
@file = __FILE__
path = File.dirname(@file)
```

But the `attr_accessor` throws a wrench into it all and cause an error to be raised, something I discovered when testing my cop on some real world code.

>  In `attr_accessor :foo`, the receiver is `nil`, but you try to enquire on the children which is probably why you get the `NoMethodError`.

So this is where I'm wondering if the code that  `def_node_search` calls should be making a `nil` check?

**Comment by thomthom** [comment]
I fixed this by changing my pattern to: 

```
(send ({lvar ivar cvar} $_) :force_encoding ...)
```
